drivers/audio/rk3xi2sbus/rk-tplg.cpp

```diff
@@ -0,0 +1,164 @@
+
+	WDF_MEMORY_DESCRIPTOR inputMemDesc;
+	WDF_MEMORY_DESCRIPTOR_INIT_BUFFER(&inputMemDesc, &inputBuffer, (ULONG)sizeof(inputBuffer));
+
+	// Send the request along
+	status = WdfIoTargetSendInternalIoctlSynchronously(
+		WdfDeviceGetIoTarget(FxDevice),
+		NULL,
+		IOCTL_ACPI_EVAL_METHOD_EX,
+		&inputMemDesc,
+		&outputSizeMemDesc,
+		NULL,
+		NULL
+	);
+
+	if (status != STATUS_BUFFER_OVERFLOW) {
+		// Method might not exist?
+		return status;
+	}
+
+	WDFMEMORY outputMemory;
+	PACPI_EVAL_OUTPUT_BUFFER outputBuffer;
+	size_t outputArgumentBufferSize = outputSizeBuffer.Length;
+	size_t outputBufferSize = FIELD_OFFSET(ACPI_EVAL_OUTPUT_BUFFER, Argument) + outputArgumentBufferSize;
+
+	WDF_OBJECT_ATTRIBUTES attributes;
+	WDF_OBJECT_ATTRIBUTES_INIT(&attributes);
+	attributes.ParentObject = FxDevice;
+
+	status = WdfMemoryCreate(&attributes,
+		NonPagedPoolNx,
+		0,
+		outputBufferSize,
+		&outputMemory,
+		(PVOID*)&outputBuffer);
+	if (!NT_SUCCESS(status)) {
+		return status;
+	}
+
+	RtlZeroMemory(outputBuffer, outputBufferSize);
+
+	WDF_MEMORY_DESCRIPTOR outputMemDesc;
+	WDF_MEMORY_DESCRIPTOR_INIT_HANDLE(&outputMemDesc, outputMemory, NULL);
+
+	status = WdfIoTargetSendInternalIoctlSynchronously(
+		WdfDeviceGetIoTarget(FxDevice),
+		NULL,
+		IOCTL_ACPI_EVAL_METHOD_EX,
+		&inputMemDesc,
+		&outputMemDesc,
+		NULL,
+		NULL
+	);
+	if (!NT_SUCCESS(status)) {
+		goto Exit;
+	}
+
+	if (outputBuffer->Signature != ACPI_EVAL_OUTPUT_BUFFER_SIGNATURE) {
+		goto Exit;
+	}
+
+	RKI2SBusPrint(DEBUG_LEVEL_ERROR, DBG_PNP,
+		"Evaluted %s successfully\n", inputBuffer.MethodName);
+
+	if (outputBuffer->Count % 2) {
+		status = STATUS_ACPI_INVALID_DATA;
+		goto Exit;
+	}
+
+	status = STATUS_NOT_FOUND;
+
+	RtlZeroMemory(rkTplg, sizeof(*rkTplg));
+	rkTplg->magic = RKTPLG_MAGIC;
+	rkTplg->length = sizeof(*rkTplg);
+
+	PACPI_METHOD_ARGUMENT currArgument = &outputBuffer->Argument[0];
+	for (ULONG i = 0; i < outputBuffer->Count; i += 2) {
+		PACPI_METHOD_ARGUMENT guidArg = currArgument;
+		currArgument = ACPI_METHOD_NEXT_ARGUMENT(currArgument);
+		PACPI_METHOD_ARGUMENT packageArg = currArgument;
+		currArgument = ACPI_METHOD_NEXT_ARGUMENT(currArgument);
+
+		if (guidArg->Type != ACPI_METHOD_ARGUMENT_BUFFER ||
+			guidArg->DataLength != 16 ||
+			packageArg->Type != ACPI_METHOD_ARGUMENT_PACKAGE) {
+			break;
+		}
+
+		if (memcmp(guidArg->Data, &GUID_ACPI_DSD, guidArg->DataLength) != 0) {
+			continue;
+		}
+
+		status = STATUS_SUCCESS;
+
+		FOR_EACH_ACPI_METHOD_ARGUMENT(dsdParameter, (PACPI_METHOD_ARGUMENT)packageArg->Data, (PACPI_METHOD_ARGUMENT)(packageArg->Data + packageArg->DataLength)) {
+			PACPI_METHOD_ARGUMENT dsdParameterName = (PACPI_METHOD_ARGUMENT)dsdParameter->Data;
+			PACPI_METHOD_ARGUMENT dsdParameterData = ACPI_METHOD_NEXT_ARGUMENT(dsdParameterName);
+
+			if (strncmp((const char*)&dsdParameterName->Data[0], "rockchip,dma", dsdParameterName->DataLength) == 0) {
```

Comment:
This will give a match for `"r"` because DataLength == 1 so it will only check the first char.

Maybe it should be something like
```
if (dsdParameterName->DataLength >= sizeof("string") &&
    0 == strcmp((const char*)&dsdParameterName->Data[0], "string"))
```

Reply:
And since you're checking DataLength, you could just use memcmp if you prefer.
```
if (dsdParameterName->DataLength >= sizeof("string") &&  
    0 == memcmp(dsdParameterName->Data, "string", sizeof("string")))
```

Reply:
Fixed